crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -232,91 +233,204 @@ impl ConsumerTask {
+                                return;
+                            }
+                        }
+                        None => {
+                            warn!("Invalid partition {partition} given to us in rebalance, ignoring it");
+                            continue;
+                        }
                     }
-
-                    // This method tells rdkafka that we have processed this message,
-                    // so its offset can be safely committed.
-                    // rdkafka periodically commits these offsets asynchronously, with a period configurable
-                    // with auto.commit.interval.ms
-                    if let Err(e) = consumer.store_offset(&topic, partition, offset) {
-                        break Err(e.into())
+                }
+            }
+            Rebalance::Revoke(partitions) => {
+                for partition in partitions.elements() {
+                    let partition = partition.into();
+                    match topic_partition_tasks.remove(&partition)
+                {
+                    Some(task_id) => {
+                        debug!("Stopping partitioned consumer for partition {partition} due to rebalance");
+                        // The partitioned queue will not be polled again.
+                        // It might be mid-poll right now, but if so its result will not be sent anywhere.
+                        if let Some(handle) = self.task_center.cancel_task(task_id) { handle.abort() }
                     }
+                    None => warn!("Kafka informed us of a revoked partition {partition} which we had no consumer task for"),
                 }
-                _ = &mut rx => {
-                    break Ok(());
+                }
+
+                match consumer.commit_consumer_state(CommitMode::Sync) {
+                    Ok(_) | Err(KafkaError::ConsumerCommit(RDKafkaErrorCode::NoOffset)) => {
+                        // Success
+                    }
+                    Err(error) => warn!("Failed to commit the current consumer state: {error}"),
                 }
             }
-        };
-        for task_id in topic_partition_tasks.into_values() {
-            self.task_center.cancel_task(task_id);
+            Rebalance::Error(_) => {}
         }
-        result
     }
 }
 
 async fn topic_partition_queue_consumption_loop(
     sender: MessageSender,
-    topic: String,
-    partition: i32,
-    topic_partition_consumer: StreamPartitionQueue<DefaultConsumerContext>,
+    partition: TopicPartition,
+    topic_partition_consumer: StreamPartitionQueue<impl ConsumerContext>,
     consumer: Arc<MessageConsumer>,
     consumer_group_id: String,
 ) -> Result<(), anyhow::Error> {
-    let mut shutdown = std::pin::pin!(cancellation_watcher());
+    debug!("Starting partitioned consumer for partition {partition}");
 
+    // this future will be aborted when the partition is no longer needed
     loop {
         tokio::select! {
```

Comment:
tokio::select! is not needed here I suppose.

Reply:
huh, good point!